src/cargo/util/toml/targets.rs

```diff
@@ -477,6 +546,98 @@ fn is_not_dotfile(entry: &DirEntry) -> bool {
     entry.file_name().to_str().map(|s| s.starts_with('.')) == Some(false)
 }
 
+fn toml_targets_and_inferred(
+    toml_targets: Option<&Vec<TomlTarget>>,
+    inferred: &[(String, PathBuf)],
+    package_root: &Path,
+    autodiscover: Option<bool>,
+    edition: Edition,
+    warnings: &mut Vec<String>,
+    target_kind_human: &str,
+    target_kind: &str,
+    autodiscover_flag_name: &str,
+) -> Vec<TomlTarget> {
+    let inferred_targets = inferred_to_toml_targets(inferred);
+    match toml_targets {
+        None => inferred_targets,
+        Some(targets) => {
+            let mut targets = targets.clone();
+
+            let target_path =
+                |target: &TomlTarget| target.path.clone().map(|p| package_root.join(p.0));
+
+            let mut seen_names = HashSet::new();
+            let mut seen_paths = HashSet::new();
+            for target in targets.iter() {
+                seen_names.insert(target.name.clone());
+                seen_paths.insert(target_path(target));
+            }
+
+            let mut rem_targets = vec![];
+            for target in inferred_targets {
+                if !seen_names.contains(&target.name) && !seen_paths.contains(&target_path(&target))
+                {
+                    rem_targets.push(target);
+                }
+            }
+
+            let autodiscover = match autodiscover {
+                Some(autodiscover) => autodiscover,
+                None => match edition {
+                    Edition::Edition2018 => true,
+                    Edition::Edition2015 => {
+                        if !rem_targets.is_empty() {
+                            let mut rem_targets_str = String::new();
+                            for t in rem_targets.iter() {
+                                if let Some(p) = t.path.clone() {
+                                    rem_targets_str.push_str(&format!("* {:?}\n", p.0))
```

Comment:
I'd recommend using `p.0.display()` here instead fo `{:?}` to avoid the extraneous quotes